doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,231 @@
+# Source Metrics for Users, v2
+
+- Associated Epic: https://github.com/MaterializeInc/materialize/issues/23154
+
+
+## The Problem
+
+The [existing source statistics](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_source_statistics_per_worker)
+we expose to users for sources answer _some important questions_, like:
+- _Generally_, how fast we are reading data from upstream?
+- Have we committed the source's snapshot?
+
+but fail to answer **two** critical questions users regularly ask about their sources:
+
+1. How much of its snapshot has this source processed?
+2. Is this source keeping up with _upstream_? (This is usually related to: can I resize this source?)
+
+
+This design document proposes metrics that answer these questions, and goes into detail about their
+implementation.
+
+## Success Criteria
+
+This design has succeeded when we are able to present these metrics to users in the console,
+and they can use them to answer questions about their source's snapshotting and sizing.
+
+## Out of Scope
+
+- Expose new metrics for sinks.
+- Solve the _hydration backpressure problem_ (more on this later)
+- Expose detailed, _granular_ metrics that are 100% accurate.
+
+## Solution Proposal
+
+### Background: _Source Frontier Differences_
+
+We define a _source frontier difference_ as the _difference_ between two source frontiers,
+as a single, numerical value. The unit of this value depends on the source frontier type.
+
+Each source implementation (critically, Kafka, Postgres, and MySQL) expose different _source timestamp_
+types. These types are partially ordered, and have an inconsistent shape. A _source frontier difference_
+is a way to capture _how different_ two frontiers are in, in a way that is consistent between source types.
+
+Source frontier differences are used in some form in the metrics proposed in [Snapshot Progress](#snapshot-progress) and
+[Steady-state](#steady-state).
+
+#### Kafka frontier differences
+
+For example, consider the following 2 kafka source frontiers:
+```
+# Partition Id: Offset
+f1: {0: 100, 1: 150}
+f2: {0: 150, 1: 170}
+```
+
+In this case, the source frontier difference is `f2 - f1 = 70 offsets`. Similarly:
+
+```
+{0: 100, 1: 100} - {0: 0} = 200 offsets
+```
+
+#### Postgres frontier differences
+
+Postgres has two separate definitions:
+
+For the number of rows in each table:
+```
+{table1: 110 rows, table2: 15 rows} - {table1: 100 rows, table2: 10 rows} = 15 rows
+```
+
+And for lsn's during replication:
+```
+100100 lsn - 100050 lsn = 50 bytes
+```
+
+#### MySQL frontier differences
+
+Similar to Postgres, mysql has a rows-per-table difference, as well as a replication difference.
+
+However, during replication, MySQL's source timestamp is partially ordered, instead of just a single
+lsn number:
+
+```
+# source id: gtid
```

Comment:
slight nit: the whole `source_id: transaction_id` format is a GTID

Reply:
ah i see